### PR TITLE
Delegate generation of authorization failure message to injected class

### PR DESCRIPTION
**Pull request by rose-a**
Allows customization of authorization error messages through dependency injection

### CODE REVIEW

**Review by Shane32 — COMMENTED** [review]
I'm not sure about this PR.  Usually one does not see interfaces/classes for generating error messages.  I would think a more typical coding pattern would be to have each of these different error messages return a specific class (that inherits `ExecutionError`), with properties for the specifics of the error, and then after the document was executed, some code can enumerate the errors and replace them as necessary with custom error messages or classes.

For example (untested, just a sample):
```cs
for (int i = 0; i <= result.Errors.Count; i++)
{
    var error = result.Errors[i];
    if (error is NotAuthorizedError notAuthError)
        error.Message = $"You are not authorized for policy '{notAuthError.PolicyName}'";
}
```

File: src/Authorization.AspNetCore/IAuthorizationFailureDescriptionGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+{
+    /// <summary>
+    /// The interface for an authorization failure description generator
+    /// </summary>
+    public interface IAuthorizationFailureDescriptionGenerator
+    {
+        /// <summary>
+        /// Generates a failure description for the given <paramref name="result"/> 
+        /// </summary>
+        /// <param name="result">the ASP.NET Core authorization result</param>
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/IAuthorizationFailureDescriptionGenerator.cs:14]:
```suggestion
        /// <param name="result">The ASP.NET Core authorization result</param>
```

File: src/Authorization.AspNetCore/IAuthorizationFailureDescriptionGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        /// <param name="operationType">the GraphQL operation type</param>
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/IAuthorizationFailureDescriptionGenerator.cs:15]:
```suggestion
        /// <param name="operationType">The GraphQL operation type. Not set if authorization occurs for the schema itself</param>
```

File: src/Authorization.AspNetCore/DefaultAuthorizationFailureDescriptionGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            foreach (var failure in result.Failure.FailedRequirements)
+                AppendFailureLine(messageBuilder, failure);
+
+            return messageBuilder.ToString();
+        }
+
+        /// <summary>
+        /// Transforms the given <paramref name="operationType"/> into a string to be used in a failure description
+        /// </summary>
+        /// <param name="operationType">the GraphQL operation type</param>
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/DefaultAuthorizationFailureDescriptionGenerator.cs:30]:
```suggestion
        /// <param name="operationType">The GraphQL operation type</param>
```

File: src/Authorization.AspNetCore/DefaultAuthorizationFailureDescriptionGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
+using System.Linq;
+using System.Text;
+using GraphQL.Language.AST;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Authorization.Infrastructure;
+
+namespace GraphQL.Server.Authorization.AspNetCore
+{
+    /// <summary>
+    /// The default authorization failure description generator for this library
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/DefaultAuthorizationFailureDescriptionGenerator.cs:10]:
```suggestion
    /// The default authorization failure description generator.
```

File: src/Authorization.AspNetCore/DefaultAuthorizationFailureDescriptionGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                OperationType.Query => "query",
+                OperationType.Mutation => "mutation",
+                OperationType.Subscription => "subscription",
+                _ => "operation",
+            };
+
+        /// <summary>
+        /// Generates a description for a single failed <see cref="IAuthorizationRequirement"/>
+        /// </summary>
+        /// <param name="messageBuilder">the message string builder</param>
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/DefaultAuthorizationFailureDescriptionGenerator.cs:44]:
```suggestion
        /// <param name="messageBuilder">The message string builder to accumulate generated text</param>
```

File: src/Authorization.AspNetCore/DefaultAuthorizationFailureDescriptionGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        /// <param name="authorizationRequirement">the failed authorization requirement</param>
+        protected virtual void AppendFailureLine(StringBuilder messageBuilder, IAuthorizationRequirement authorizationRequirement)
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/DefaultAuthorizationFailureDescriptionGenerator.cs:46]: messageBuilder -> error

**Comment by sungam3r** [comment]
изображение

**Comment by sungam3r** [comment]
> I would think a more typical coding pattern would be to have each of these different error messages return a specific class (that inherits ExecutionError), with properties for the specifics of the error, and then after the document was executed, some code can enumerate the errors and replace them as necessary with custom error messages or classes.

This solution can be applied, but it is more difficult to maintain. In addition, we already have ExecutionError with a very **definite meaning**, we do not need **new types** of errors, the whole question is to get an auxiliary description.

**Comment by sungam3r** [comment]
@Shane32 My previous comment was wrong. Indeed, it is possible to make separate errors instead of _generic_ one. It might have been worth doing right from the start of project, but now it would be a backward incompatible change and would require a clients change. I like the idea with separate errors. @rose-a ?

**Comment by Shane32** [comment]
Another idea would be to simply add protected methods to the `AuthorizationValidationRule` class to generate the respective error messages.  If desired, one can simply replace the validation rule with their own implementation, overriding the members required.

I also drafted a sample of a non-breaking change to the existing codebase that demonstrates moving all the error message generation code into a new `AuthorizationError` class.  Unfortunately `Exception.Message` is a read-only property so it is not easy to override a message once the class has been constructed.  However, by making `AuthorizationValidationRule.AddValidationError` protected, it is easy to replace the entire validation rule with an implementation that creates a custom derivative of `AuthorizationError`.  This can be further separated by having the code within `AddValidationError` create a specific class (deriving from `AuthorizationError`) depending on which exact type of `IAuthorizationRequirement` failed.

I suggest we also add a PR to the main project that adds a `SetMessage(string message)` method to `ExecutionError`.  This would allow for an easier way to adjust messages after the class has been created.

Please keep in mind that I don't actually use the server project, and so you both are probably more familiar with how this would work in practice @sungam3r @rose-a .

**Comment by sungam3r** [comment]
> I suggest we also add a PR to the main project that adds a SetMessage(string message) method to ExecutionError. This would allow for an easier way to adjust messages after the class has been created.

I would not do that.

**Comment by Shane32** [comment]
> > I suggest we also add a PR to the main project that adds a SetMessage(string message) method to ExecutionError. This would allow for an easier way to adjust messages after the class has been created.
> 
> I would not do that.

By that logic, `AddLocation` and `Code { set; }` and `Number { set; }` should be `protected`, not `public`.  But if everything is writable, then `Message` should be as well.

**Comment by rose-a** [comment]
I'm ok with creating "error classes" like @Shane32 started in #479... But I'd go further and make `AuthorizationError` abstract and create a `MissingClaimsAuthorizationError`, `NameAuthorizationError`, ... which provide the appropriate message.

I'd still like to put the switch logic in a factory class to be able to inject a custom version which returns customized errors in certain cases...

Another way would be to put the error generation for each case in a separate virtual method to be able to easily override just one of them and leave the switch logic alone...

**Comment by rose-a** [comment]
> 
> 
> > > I suggest we also add a PR to the main project that adds a SetMessage(string message) method to ExecutionError. This would allow for an easier way to adjust messages after the class has been created.
> > 
> > 
> > I would not do that.
> 
> By that logic, `AddLocation` and `Code { set; }` and `Number { set; }` should be `protected`, not `public`. But if everything is writable, then `Message` should be as well.

I think that's because the `Message` property of `Exception` is not writeable...

**Comment by rose-a** [comment]
> I'm not sure about this PR. Usually one does not see interfaces/classes for generating error messages. 

Well, it was the simplest way to create a custom error message for my project 😉

**Comment by Shane32** [comment]
> I'm ok with creating "error classes" like @Shane32 started in #479... But I'd go further and make `AuthorizationError` abstract and create a `MissingClaimsAuthorizationError`, `NameAuthorizationError`, ... which provide the appropriate message.

I agree

**Comment by Shane32** [comment]
> Another way would be to put the error generation for each case in a separate virtual method to be able to easily override just one of them and leave the switch logic alone...

I like that also

**Comment by sungam3r** [comment]
> But if everything is writable, then Message should be as well.

This is a consequence of inheritance from `System.Exception`. We should to take it as it is.

**Comment by sungam3r** [comment]
> But I'd go further and make AuthorizationError abstract and create a MissingClaimsAuthorizationError, NameAuthorizationError, ... which provide the appropriate message.

Of course it should be done.

File: src/Authorization.AspNetCore/AuthorizationValidationRule.cs [hunk 7]
```diff
@@ -20,16 +17,19 @@ public class AuthorizationValidationRule : IValidationRule
...
         private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly IAuthorizationFailureDescriptionGenerator _failureDescriptionGenerator;
 
         /// <summary>
         /// Creates an instance of <see cref="AuthorizationValidationRule"/>.
         /// </summary>
         /// <param name="authorizationService"> ASP.NET Core <see cref="IAuthorizationService"/> to authorize against. </param>
         /// <param name="httpContextAccessor"> ASP.NET Core <see cref="IHttpContextAccessor"/> to take user (HttpContext.User) from. </param>
-        public AuthorizationValidationRule(IAuthorizationService authorizationService, IHttpContextAccessor httpContextAccessor)
+        /// <param name="failureDescriptionGenerator"> The generator for authorization failure description messages</param>
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/AuthorizationValidationRule.cs:27]:
```suggestion
        /// <param name="failureDescriptionGenerator"> The generator for authorization failure description messages. </param>
```

**Comment by Shane32** [comment]
> > But I'd go further and make AuthorizationError abstract and create a MissingClaimsAuthorizationError, NameAuthorizationError, ... which provide the appropriate message.
> 
> Of course it should be done.

Once completed, an easy solution to the original problem is to provide an `IErrorInfoProvider` implementation like this:

```cs
    public class MyErrorInfoProvider : ErrorInfoProvider
    {
        public override ErrorInfo GetInfo(ExecutionError executionError)
        {
            var info = base.GetInfo(executionError);
            info.Message = executionError switch
            {
                SyntaxError syntaxError => "Syntax error. READY",
                ValidationError validationError => $"I'm afraid I can't do that, Dave.  You've got a {validationError.Number}",
                _ => info.Message,
            };
            return info;
        }
    }
```
Injecting an `IErrorInfoProvider` is already built into the framework.

**Comment by sungam3r** [comment]
So as you demonstrated there is no need to have setter for `ExecutionError.Message` since we can return arbitrary text via `ErrorInfo`.

**Comment by rose-a** [comment]
What do we do about `InternalErrorInfoProvider` inside `GraphQLBuilderCoreExtensions`?

**Comment by rose-a** [comment]
After playing around a little bit i don't think it makes sense to make separate classes for the different authorization errors... since one `AuthorizationError` can contain multiple failed requirements...

I`m willing close this PR in favour of #479...

**Comment by sungam3r** [comment]
OK. You may branch from #479 and move on.

**Comment by rose-a** [comment]
closing this in favour of #480

**Pull request closed without merging**